subsys/net/ip/tcp.c

```diff
@@ -997,6 +1020,11 @@ static int tcp_unsent_len(struct tcp *conn)
 	}
 
 	unsent_len = conn->send_data_total - conn->unacked_len;
+	if (conn->unacked_len >= conn->send_win) {
```

Comment:
Generally, the concern I have regarding this change, is that we modify the `tcp_unsent_len()` function content, so that it does not always return true - as it now takes the remaining window size into account as well. So we don't really return the amount of data that we should send, but rather the amount we can. I find this a little confusing, but it might be just me.

Given that, I wonder if it wouldn't be a better approach to leave this function intact, and rather modify `send_data_total ` value only after successful send - that way `tcp_window_full()` could be used to retrieve the window status regardless of the context it is called in.

Reply:
> I just looked at the function `net_tcp_queue_data`. When an -ENOBUFS by `tcp_send_queued_data` is returned, it throws away the whole block of size `len` from the `send_data`. If the `len` is > MSS, it could happen that the first section is transmitted, but at the second an -ENOBUFS occurs. In that case the data is transmitted, but later on removed from the send_data. What this will do to the sequence number administration is a big question.

Just added a possible fix for this potential issue.